active/InputValidation/InputValidation.md

```diff
@@ -0,0 +1,572 @@
+Use INotifyDataErrorInfo with x:Bind to update your view. This API can be implemented by any class, and allows the object to raise an event if one of its properties is set to an invalid value. When the source of a Xaml x:Bind implements this interface, the x:Bind will put the validation errors onto the bind target. Xaml controls such as TextBox will recognize these errors and provide visualization for them.
+
+### XAML
+
+a user input control with the input bound:
+
+```xaml
+<TextBox Name="ValidationTextbox" 
+         Text="{x:Bind  ViewModel.Person.Name} "/>
+
+```
+#### Code - ViewModel
+
+The “ViewModel” property is the View Model for the Page on which this TextBox is displayed (this.ViewModel).
+``` csharp
+
+public class ViewModel
+{
+    public ViewModel()
+    {
+        Person = new Person();
+    }
+
+    public Person Person { get; set; }
+
+}
+```
+
+#### C# only - Person (Model)
+Person.Name is the source of the bind (this.ViewModel.Person.Name).  Person derives from ValidationWrapper (app code) which will provide the data validation for Name.  It’s marked with [MinLength] and [MaxLength] data annotations.  The property accessors use GetValue/SetValue in the base ValidationWrapper class for storage and to check the validity.
+
+``` csharp
+using System.ComponentModel.DataAnnotations;
+    
+public class Person : ValidationWrapper
+{
+    [MinLength(4, ErrorMessage = "Name must be at least 4 characters")]
+    [MaxLength(6, ErrorMessage = "Name must be no more than 6 characters")]
+    public string Name
+    {	
+        get { return GetValue<string>("Name"); }
+        set { SetValue("Name", value); }
+    }
+}
+```
+
+#### Code – INotifyDataErrorInfo base class
+The NotifyDataErrorInfoBase class (app code) provides a reusable implementation of INotifyDataErrorInfo (new API).  It’s the base of ValidationWrapper, which is the base class for Person.
+``` csharp
+public class NotifyDataErrorInfoBase : INotifyDataErrorInfo
+{
+    protected readonly Dictionary<string, ObservableCollection<ValidationError>> Errors;
+
+    protected NotifyDataErrorInfoBase()
+    {
+        Errors = new Dictionary<string, List<string>>();
+    }
+
+    public bool HasErrors => Errors.Any();
+
+    public event EventHandler<DataErrorsChangedEventArgs> ErrorsChanged;
+
+    public ObservableCollection<ValidationError> GetErrors(string propertyName)
+    {
+        return propertyName != null && Errors.ContainsKey(propertyName)
+            ? Errors[propertyName]
+            : new ObservableCollection<ValidationError>();
+    }
+
+    protected virtual void OnErrorsChanged(string propertyName)
+    {
+        ErrorsChanged?.Invoke(this, new DataErrorsChangedEventArgs(propertyName));
+    }
+
+    protected void ClearErrors()
+    {
+        foreach (var propertyName in Errors.Keys.ToList())
+        {
+            Errors.Remove(propertyName);
+            OnErrorsChanged(propertyName);
+        }
+    }
+}
+```
+#### Code – ValidationWrapper (no new APIs)
```

Comment:
Again, is this what will be provided? or is it an example of something a developer would write themselves? It seems generic enough that it could be used by anyone and not need recreating by each developer.

Reply:
Thanks for the feedback. Yes this is boilerplate code that the developer would have to add. What if we added this as part of the community toolkit to make it easier?

Reply:
Toolkit seems like a good place for this